### PR TITLE
Implement tmt status command

### PR DESCRIPTION
**Pull request by FrNecas**
I am creating this draft PR to gather some feedback on the functionality proposed in #416 . All of the options are implemented so the output can be filtered using options `--active`, `--abandoned`, `--finished`. 

One of the aspects I am not completely sure about is how the command should work after calling for example `tmt run provision` and then `tmt status`. The verbose status of the run looks like this:
```
disc prov prep exec repo fini  id
todo done todo todo todo todo  /var/tmp/tmt/run-004  /test
```
The current implementation shows `discover` as the overall status of the run/plan when running only with `-v` or with no verbosity because it is the first step that hasn't been done. Would it perhaps make more sense to use `prepare` which is the next step after the last done step?

When discussing this, @lukaszachy also suggested that it would perhaps be good to introduce new status - running for steps that are in progress but haven't finished yet (currently it's either todo or done). What do you think about this @psss ? Also, is there a way how to check if a directory is a valid run directory? I haven't been able to figure out how to do so.

### CODE REVIEW

**Comment by psss** [comment]
> The current implementation shows `discover` as the overall status of the run/plan when running only with `-v` or with no verbosity because it is the first step that hasn't been done. Would it perhaps make more sense to use `prepare` which is the next step after the last done step?

This is an interesting question. There are quite many different ways how to select which steps should be run. Here are some common examples:

    tmt run
    tmt run discover
    tmt run provision finish login

And we have different expectations for these use cases. Run all steps for the first one, run only `discover` and nothing more for the second one and finally cherry-pick only desired steps for the last example.

Thinking about it a bit more, here's a little bit different approach: What about showing the last completed step instead? This would give a clear idea about what has been already finished. And works relatively nicely for the above-mentioned examples as well.

Although some additional information, such as list of selected steps or enabled plans, could be extracted from the `run.yaml` file in the root of the run directory, we still cannot guess what the user will do in the next step. For `tmt run --until execute` does not mean that `report` and `finish` steps won't be ever executed.

**Comment by psss** [comment]
> When discussing this, @lukaszachy also suggested that it would perhaps be good to introduce new status - running for steps that are in progress but haven't finished yet (currently it's either todo or done).

Unless necessary I would like to prevent adding a new status because the implementation which takes care of checking what has been already done and re-running unfinished steps is already rather complex. Not saying it would be impossible but as for now I'm not sure the benefit would be worth it.

**Comment by psss** [comment]
> Also, is there a way how to check if a directory is a valid run directory?

I think a good approach could be to check for the presence of the `run.yaml` file which should always be created in the root of the run work directory.

@FrNecas, and thanks much for working on this! Looks very nice already now :-) One idea for improvement: I'd suggest to use color for all statuses so that the whole `status` column is in color.

**Comment by psss** [comment]
I ran into the following issue:

    tmt run provision finish login

Then during the interactive session I run in the second terminal:

    tmt status
    done       /var/tmp/tmt/run-013

And when running in verbose mode the run is not displayed at all:

    tmt status -v
    <nothing>

This might be caused by the fact that `/plans/default` is used in the case.

**Comment by FrNecas** [comment]
@psss thanks for the feedback, good suggestion with showing the last step. You are right that we can't really guess what's coming next so showing the last thing done seems like a better approach. I will add the `run.yaml` check that should be good enough. I will take a look at what's happening with the login example :)

**Comment by FrNecas** [comment]
Updated the PR based on the given feedback
- tmt status now shows the latest done step
- it also checks run.yaml existence before examining the directory
- the whole status column is now colored (I've chosen green for done and yellow for anything else that should be clear)
- @psss I didn't manage to reproduce the issue after the aforementioned changes. Would you mind trying again? I see you were using testcloud provision, it may have been caused by not checking for `run.yaml`. I remember seeing /var/tmp/tmt/testcloud directory being checked by tmt status which obivously isn't a valid run directory... Running tmt status during interactive shell seems to be working fine right now.

**Comment by psss** [comment]
Thanks for all the improvements. Regarding the testcloud provision: Sorry, that was fault on my side, did a bit too much of cleanup during `tmt status` experimenting and forgot about it :)

However, the issue with verbose output missing some runs seems to be still present. Try `tmt run provision login finish` in one terminal and then `tmt status` in another one. It says `done` although the interactive session is still running. And for verbose output (both `-v` and `-vv`) it completely omits that run. Are you able to reproduce?

File: tmt/cli.py [hunk 1]
```diff
@@ -832,3 +832,40 @@ def init(context, path, template, force, **kwargs):
...
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+@main.command()
+@click.pass_context
+@click.argument('path', default=tmt.utils.WORKDIR_ROOT)
+@click.option(
+    '-i', '--id', 'id_', help='Run id (name or directory path) to show status of.', metavar="ID")
+@click.option(
+    '--abandoned', is_flag=True, default=False,
+    help='List runs with active provision.')
```
Comment (psss) [inline tmt/cli.py:848]: Perhaps we could make it a bit more clear what `active` actually means. What about something like "runs which have the `provision` step completed but `finish` not yet done"?

File: tmt/cli.py [hunk 2]
```diff
@@ -832,3 +832,40 @@ def init(context, path, template, force, **kwargs):
...
+@click.option(
+    '--active', is_flag=True, default=False,
+    help='List runs where at least one step has not been finished.')
```
Comment (psss) [inline tmt/cli.py:851]: Here it should probably be "at least one of the enabled steps has not been finished".

File: tmt/cli.py [hunk 3]
```diff
@@ -832,3 +832,40 @@ def init(context, path, template, force, **kwargs):
...
+@click.option(
+    '--finished', is_flag=True, default=False,
+    help='List all finished runs.')
```
Comment (psss) [inline tmt/cli.py:854]: This should be probably something like "runs which have all enabled steps completed".

File: tmt/cli.py [hunk 4]
```diff
@@ -832,3 +832,40 @@ def init(context, path, template, force, **kwargs):
...
+@verbose_debug_quiet
+def status(context, id_, path, abandoned, active, finished, **kwargs):
```
Comment (psss) [inline tmt/cli.py:856]: Hint: If you use `Status._save_context()` you don't have to pass individual options and can easily check them using `self.opt()` later.

**Review by psss — CHANGES_REQUESTED** [review]
A couple of enhancements proposed (haven't reviewed the `Status` class yet).

**Comment by FrNecas** [comment]
Updated the options, replaced unnecessary passing of the filtering options with self.opt() calls. I was unable to reproduce your issue :/ I am using a very simple test plan
```
summary: Basic smoke test
execute:
    script: bash --help
```

Running `tmt run provision login finish` and then `tmt status`:
```
status     id
provision  /var/tmp/tmt/run-001
```
Running `tmt status -v`
```
status     id
provision  /var/tmp/tmt/run-001  /test
```

Running `tmt status -vv`
```
disc prov prep exec repo fini  id
todo done todo todo todo todo  /var/tmp/tmt/run-001  /test
```

Exiting the interactive session and running `tmt status`:
```
status     id
done       /var/tmp/tmt/run-001
```

This seems to work as expected

**Comment by psss** [comment]
Thanks for simplifying the option handling.

> I was unable to reproduce your issue :/ I am using a very simple test plan

I guess the problem is caused (as already mentioned above) by the fact that the `/plans/default` is used. Have you tried to run `tmt run provision login finish` in a directory with no `fmf` metadata?

**Comment by FrNecas** [comment]
I managed to reproduce the issue and discovered that it is related to how default plan is handled when loading a run. `Run._save_tree` would read the local fmf metadata which could be incorrect (relative to the run being loaded). After discussion with @psss , we can detect whether the default plan should be used based on the `root` field (use default plan if root isn't defined). The status seems to be correct now, @psss could you check on your end as well please?

**Comment by psss** [comment]
Nice, thanks for fixing this. I've tested the `/default/plan` scenario and can confirm the traceback is gone. I did some more experimenting and ran into two issues: Picking specific run does not seem to work:

```
> tmt status -i run-003
status     id
done       /var/tmp/tmt/run-001
done       /var/tmp/tmt/run-002
done       /var/tmp/tmt/run-003
```

And second, when `tmt status` is used when there is a run being executed it causes the run to crash. This is caused by `Step.wake()` method which [cleans up](https://github.com/psss/tmt/blob/master/tmt/steps/__init__.py#L126) the workdir. Here's the relevant part of the traceback for reference:

```python
  File "/home/psss/git/tmt/tmt/cli.py", line 877, in status
    status_obj.show()
  File "/home/psss/git/tmt/tmt/base.py", line 1180, in show
    self.process_run(run)
  File "/home/psss/git/tmt/tmt/base.py", line 1144, in process_run
    step.wake()
  File "/home/psss/git/tmt/tmt/steps/discover/__init__.py", line 75, in wake
    super(Discover, self).wake()
  File "/home/psss/git/tmt/tmt/steps/__init__.py", line 131, in wake
    self._workdir_cleanup()
  File "/home/psss/git/tmt/tmt/utils.py", line 480, in _workdir_cleanup
    shutil.rmtree(directory)
```
It seems we will need to implement something like a `read-only` wake up.

File: tmt/base.py [hunk 5]
```diff
@@ -1032,6 +1042,144 @@ def go(self):
...
+        for i in range(len(steps) - 1, -1, -1):
+            if steps[i].status() == 'done':
+                if i + 1 == len(steps):
+                    # Last enabled step, consider the whole plan done
+                    return 'done'
+                else:
+                    return step_names[i]
+        return 'todo'
+
+    def plan_matches_filters(self, plan):
```
Comment (psss) [inline tmt/base.py:1065]: Please, include a docstring. Similarly below. Thanks.

File: tmt/cli.py [hunk 6]
```diff
@@ -835,3 +835,43 @@ def init(context, path, template, force, **kwargs):
...
+@click.option(
+    '--finished', is_flag=True, default=False,
+    help='List all runs which have all enabled steps completed.')
+@verbose_debug_quiet
+def status(context, id_, path, abandoned, active, finished, **kwargs):
+    """
+    Show status of runs.
+
+    Lists past runs in the given directory filtered using options.
+    /var/tmp/tmt is used by default.
```
Comment (psss) [inline tmt/cli.py:867]: It would be good to mention what exactly does the `done` status mean. Also at least a short description of the verbose levels would be useful here.

**Review by psss — CHANGES_REQUESTED** [review]
A couple of docs issues. Also it would be good to add at lest some basic test coverage.

File: tmt/cli.py (hunk 6 above)
Comment (FrNecas) [reply tmt/cli.py:867]: I am not completely sure how would I go about describing `done` status. Do you have any suggestions?

**Comment by FrNecas** [comment]
I've fixed most of the things that you pointed out (id not working was caused by a bad `self.opt()` call; I've added docstrings), thanks for the review @psss . I've replaced `Step.wake()` with `Step.load()` which only reads the metadata of the step which should be sufficient for tmt status (the status of the step is loaded but it doesn't clean the workdir).

I will hopefully add some basic test coverage for all the various options tomorrow.

File: tmt/base.py [hunk 7]
```diff
@@ -1032,6 +1042,151 @@ def go(self):
...
+        for plan in run.plans:
+            if self.plan_matches_filters(plan):
+                for step in plan.steps(disabled=True):
+                    column = step.status() + ' '
+                    echo(self.colorize_column(column), nl=False)
+                echo(f' {run.workdir}  {plan.name}')
+
+    def process_run(self, run):
+        """ Display the status of the given run based on verbosity """
+        run.load()
```
Comment (psss) [inline tmt/base.py:1148]: It would be good to handle possible `GeneralError` exception here becase the metadata tree root might not be always accessible. In that case the first such run causes this error:

    Invalid yaml syntax: Invalid directory path: /home/psss/git/tmt/tests/test/select/data

File: tmt/cli.py (hunk 6 above)
Comment (psss) [reply tmt/cli.py:867]: I meant something like this:

    By default, the last completed step is displayed,
    `done` is used when all enabled steps are completed.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for all the changes. Looks very good! I just encountered one weird issue:

    > rm -rf /var/tmp/tmt/run-*
    > tmt run -q --until execute plan -n core test -n smoke
    > tmt status
    status     id
    done       /var/tmp/tmt/run-001

    > rm -rf /var/tmp/tmt/run-*
    > tmt run -q plan -n core test -n smoke
    > tmt run -q --until execute plan -n core test -n smoke
    > tmt status
    status     id
    done       /var/tmp/tmt/run-001
    execute    /var/tmp/tmt/run-002

The `--until execute` run should be `done` in the second case as well as all enabled steps have been completed.

File: tmt/base.py (hunk 7 above)
Comment (psss) [reply tmt/base.py:1148]: Perhaps something like this?

```python
try:
    run.load()
except tmt.utils.GeneralError as error:
    self.warn(f'Failed to check {run.workdir} ({error}).')
    return
```

**Review by psss — APPROVED** [review]
Thanks for addressing the last remaining issues (we'll fix that hardly reproducible corner case later). I've just fixed a minor traceback which probably happens for interrupted runs (no status defined). Rebased on the latest master. Let's see what tests will say...

**Comment by FrNecas** [comment]
I am working on adding some tests for the basic use-cases. Will hopefully be able to finish it tomorrow so lets wait a bit with merging this one

**Comment by psss** [comment]
Sure, no problem.

**Comment by FrNecas** [comment]
@psss The weird issue should be gone. I did a workaround that revolves around making a copy of the common object stored in context to stop a run from modifying the context. It is far from perfect and it would be better to somehow fix this one in Run/Common but I think we can improve this further down the line (it would probably require a fair bit of refactoring which isn't relevant to this PR).

I've also added some test coverage but I didn't manage to make a test for `--active` option. Is there a way to enable all steps in tmt but only run some of them?

**Comment by psss** [comment]
> I've also added some test coverage but I didn't manage to make a test for `--active` option. Is there a way to enable all steps in tmt but only run some of them?

I've covered that in d737f95 by making the `prepare` step to fail.
